### PR TITLE
Fix static_casts leading to overflows.

### LINKED ISSUE #644: CUDA Tests broken in Develop
What are the exact, minimal CMake parameters to run the tests *manually* with CUDA enabled?

I am trying
```bash
mkdir build
cd build

cmake .. -DALPAKA_ACC_GPU_CUDA_ENABLE=ON -DALPAKA_CUDA_ARCH="37"
# ok

make -j
# ok

ctest -V
# fail 9/26
```

on a K80 but many tests fail:
```
65% tests passed, 9 tests failed out of 26

Total Test time (real) = 112.39 sec

The following tests FAILED:
	 12 - atomic (Failed)
	 13 - blockShared (Failed)
	 14 - blockSync (Failed)
	 18 - kernel (Failed)
	 19 - memBuf (Failed)
	 20 - memView (Failed)
	 21 - memP2P (Failed)
	 24 - rand (Failed)
	 25 - time (Failed)
Errors while running CTest
make: *** [test] Error 8
```

e.g. `rand`:
```
$ ./test/unit/rand/rand 
Running 18 test cases...
rand: /home/huebl/src/alpaka/include/alpaka/workdiv/WorkDivHelpers.hpp:192: alpaka::workdiv::WorkDivMembers<TDim, TIdx> alpaka::workdiv::subDivideGridElems(const alpaka::vec::Vec<TDim, TVal>&, alpaka::vec::Vec<TDim, TVal>, const alpaka::acc::AccDevProps<TDim, TIdx>&, bool, alpaka::workdiv::GridBlockExtentSubDivRestrictions) [with TDim = std::integral_constant<long unsigned int, 1ul>; TIdx = short int]: Assertion `isValidAccDevProps(accDevProps)' failed.
unknown location(0): fatal error: in "rand_/defaultRandomGeneratorIsWorking<N6alpaka3acc12AccGpuCudaRtISt17integral_constantImLm1EEsEE>": signal: SIGABRT (application abort requested)
/home/huebl/src/alpaka/test/unit/rand/src//RandTest.cpp(152): last checkpoint: "defaultRandomGeneratorIsWorking" entry.
rand: /home/huebl/src/alpaka/include/alpaka/workdiv/WorkDivHelpers.hpp:192: alpaka::workdiv::WorkDivMembers<TDim, TIdx> alpaka::workdiv::subDivideGridElems(const alpaka::vec::Vec<TDim, TVal>&, alpaka::vec::Vec<TDim, TVal>, const alpaka::acc::AccDevProps<TDim, TIdx>&, bool, alpaka::workdiv::GridBlockExtentSubDivRestrictions) [with TDim = std::integral_constant<long unsigned int, 2ul>; TIdx = short int]: Assertion `isValidAccDevProps(accDevProps)' failed.
unknown location(0): fatal error: in "rand_/defaultRandomGeneratorIsWorking<N6alpaka3acc12AccGpuCudaRtISt17integral_constantImLm2EEsEE>": signal: SIGABRT (application abort requested)
/home/huebl/src/alpaka/test/unit/rand/src//RandTest.cpp(152): last checkpoint: "defaultRandomGeneratorIsWorking" entry.
rand: /home/huebl/src/alpaka/include/alpaka/workdiv/WorkDivHelpers.hpp:192: alpaka::workdiv::WorkDivMembers<TDim, TIdx> alpaka::workdiv::subDivideGridElems(const alpaka::vec::Vec<TDim, TVal>&, alpaka::vec::Vec<TDim, TVal>, const alpaka::acc::AccDevProps<TDim, TIdx>&, bool, alpaka::workdiv::GridBlockExtentSubDivRestrictions) [with TDim = std::integral_constant<long unsigned int, 3ul>; TIdx = short int]: Assertion `isValidAccDevProps(accDevProps)' failed.
unknown location(0): fatal error: in "rand_/defaultRandomGeneratorIsWorking<N6alpaka3acc12AccGpuCudaRtISt17integral_constantImLm3EEsEE>": signal: SIGABRT (application abort requested)
/home/huebl/src/alpaka/test/unit/rand/src//RandTest.cpp(152): last checkpoint: "defaultRandomGeneratorIsWorking" entry.

*** 3 failures are detected in the test module "rand"
```

with CUDA 8.0.44, Boost 1.62 on Hypnos (HZDR) and `develop` as of 8e206bdad75da2704bdac7060a61115a97304df2

### PR DESCRIPTION
**Pull request by BenjaminW3**
Fixes #644

When the `Idx` type is too small (e.g. short in this case), the maximum grid and thread sizes may have overflown leading to invalid values. This resulted in assertions which failed.
This should not have been a problem in any real world application because most users use larger index types. We may want to review all other `static_cast`s :-)

Tasks
* [x] Fix compilation on non-MSVC compilers
* [x] Add unit tests for `saturateCast`, `[Lower/Higher][Min/Max]`
* [x] @ax3l @psychocoderHPC  Test if this helps

### CODE REVIEW

**Comment by BenjaminW3** [comment]
I broke something with my latest changes. Please wait before testing this.

**Comment by BenjaminW3** [comment]
Should be fixed now. Ready for testing.

**Comment by ax3l** [comment]
@BenjaminW3 already fixes most of the tests for me. thanks a lot! :)

Still failing are:
```
The following tests FAILED:
	 15 - memBuf (Failed)
	 16 - memView (Failed)
	 17 - memP2P (Failed)
Errors while running CTest
```

with
```
test/common/include/alpaka/test/mem/view/ViewTest.hpp(125): fatal error: in "memView/viewSubViewOffsetConstTest<N6alpaka3acc12AccGpuCudaRtISt17integral_constantImLm3EEsEE>": critical check pitchView[i-1] >= pitchMinimum[i-1] has failed [-24576 < 768]

ALPAKA_CHECK failed because '!(pBytes[i] == byte)'
test/common/include/alpaka/test/mem/view/ViewTest.hpp(225): fatal error: in "memP2P/memP2PTest<N6alpaka3acc12AccGpuCudaRtISt17integral_constantImLm3EEsEE>": critical check true == fixture( verifyBytesSet, alpaka::test::mem::view::begin(view), alpaka::test::mem::view::end(view), byte) has failed [true != false]
```

**Comment by BenjaminW3** [comment]
The other failures seem to be independent of this error. I would like to merge this as soon as possible.

**Comment by ax3l** [comment]
The other errors also look like a `short` overflow to me.

**Comment by BenjaminW3** [comment]
Even if this is the case I would handle them separately. This PR fixes a bug in alpaka. The other errors are most likely bugs in the tests due to not carefully chosen values.

**Comment by ax3l** [comment]
I'll more this for now.

File: include/alpaka/core/SaturateCast.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+namespace alpaka
+{
+    namespace core
+    {
+        //-----------------------------------------------------------------------------
+        //! \return The input casted and saturated to T.
+        template<
+            typename T,
+            typename V>
+        auto saturateCast(
```
Comment (ax3l) [inline include/alpaka/core/SaturateCast.hpp]: Actually, saturated is a really weird naming. You mean "clipped", don't you?

**Comment by ax3l** [comment]
@BenjaminW3 ranges of axes, colorbars, value ranges are "clipped", not saturated.

**Comment by ax3l** [comment]
I am also not surprised the low-level CUDA intrinsic calls this "saturate" since CUDA is also heavily image influenced.

File: include/alpaka/core/SaturateCast.hpp (hunk 1 above)
Comment (ax3l) [reply include/alpaka/core/SaturateCast.hpp]: suggestion: `clipCast`

File: include/alpaka/core/SaturateCast.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <algorithm>
+#include <limits>
+
+namespace alpaka
+{
+    namespace core
+    {
+        //-----------------------------------------------------------------------------
+        //! \return The input casted and saturated to T.
```
Comment (ax3l) [inline include/alpaka/core/SaturateCast.hpp]: suggestion: improve details.
Write "and clipped to the maximum allowed range of T"

**Comment by ax3l** [comment]
I see, did not find the CUDA file (CUDA... too close to shaders ;-) ). And [this wikipedia article](https://en.wikipedia.org/wiki/Saturation_arithmetic). Nevertheless, the term is DSP and image-motivated which is also the background for naming corresponding intrinsics as such.

**Pull request merged**